snapshotter/devmapper/pool_device.go

```diff
@@ -0,0 +1,297 @@
+type PoolDevice struct {
+	poolName        string
+	currentDeviceID int
+	devices         map[string]int
+}
+
+// NewPoolDevice creates new thin-pool from existing data and metadata volumes.
+// If pool 'poolName' already exists, it'll be reloaded with new parameters.
+func NewPoolDevice(ctx context.Context, poolName, dataVolume, metaVolume string, blockSizeSectors uint32) (*PoolDevice, error) {
+	log.G(ctx).Infof("initializing pool device '%s'", poolName)
+
+	driverVersion, err := devicemapper.GetDriverVersion()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get driver version")
+	}
+
+	log.G(ctx).Debugf("using driver: %s", driverVersion)
+
+	libVersion, err := devicemapper.GetLibraryVersion()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get library version")
+	}
+
+	log.G(ctx).Debugf("using lib version: %s", libVersion)
+
+	dataFile, err := os.Open(dataVolume)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open data volume")
+	}
+
+	defer dataFile.Close()
+
+	metaFile, err := os.Open(metaVolume)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open meta volume")
+	}
+
+	defer metaFile.Close()
+
+	poolPath := getDevMapperPath(poolName)
+	if _, err := os.Stat(poolPath); err == nil {
+		log.G(ctx).Debugf("reloading existing pool '%s'", poolPath)
+		if err := devicemapper.ReloadPool(poolName, dataFile, metaFile, blockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to reload pool '%s'", poolName)
+		}
+	} else {
+		if !os.IsNotExist(err) {
+			return nil, errors.Wrapf(err, "failed to stat for '%s'", poolPath)
+		}
+
+		log.G(ctx).Debug("creating new pool device")
+		if err := devicemapper.CreatePool(poolName, dataFile, metaFile, blockSizeSectors); err != nil {
+			return nil, errors.Wrapf(err, "failed to create thin-pool with name '%s'", poolName)
+		}
+	}
+
+	return &PoolDevice{
+		poolName: poolName,
+		devices:  make(map[string]int),
+	}, nil
+}
+
+func (p *PoolDevice) CreateThinDevice(deviceName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if _, ok := p.devices[deviceName]; ok {
+		return 0, errors.Errorf("device with name '%s' already created", deviceName)
+	}
+
+	// Create device, retry if device id is taken
+	deviceID, err := p.tryAcquireDeviceID(func(thinDeviceID int) error {
+		return devicemapper.CreateDevice(p.poolName, thinDeviceID)
+	})
+
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to create thin device")
+	}
+
+	p.devices[deviceName] = deviceID
+
+	devicePath := p.GetDevicePath(p.poolName)
+	if err := devicemapper.ActivateDevice(devicePath, deviceName, deviceID, virtualSizeBytes); err != nil {
+		return 0, errors.Wrap(err, "failed to activate thin device")
+	}
+
+	return deviceID, nil
+}
+
+func (p *PoolDevice) CreateSnapshotDevice(deviceName string, snapshotName string, virtualSizeBytes uint64) (int, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	deviceID, ok := p.devices[deviceName]
```

Comment:
I'm guessing `deviceName` never equals `snapshotName`? Or it'll always run into one of these errors.

Reply:
They shouldn't be equal. You mean missing check?

Reply:
I don't think we need a check, but I was curious if they could be equal.